ContextImpl can silently ignore Exceptions

The test class at the end of this comment doesn't report the fatal StackOverflowError at all - it just silently swallows it.  If the call to make RxJava use the vertx scheduler is skipped - i.e. line 23 is commented out - then the exception is reported on the console. 

The output from the code using vertx schedulers is...

```
Starting...
vertxHandler(): Entry
Sleeping...
throwFatal(hello)
...Awake
vertxHandler(): Exit
...Done
```

and using the RxSchedulers (i.e. with L23 commented out) it's...

```
Starting...
vertxHandler(): Entry
Sleeping...
throwFatal(hello)
Rx error handler invoked
io.reactivex.exceptions.UndeliverableException: java.lang.StackOverflowError: FATAL!!
	at io.reactivex.plugins.RxJavaPlugins.onError(RxJavaPlugins.java:349)
	at io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:69)
	at io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:57)
	at java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:266)
	at java.util.concurrent.FutureTask.run(FutureTask.java)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.StackOverflowError: FATAL!!
	at FatalSwallowed.throwFatal(FatalSwallowed.java:71)
	at FatalSwallowed.lambda$3(FatalSwallowed.java:60)
	at io.reactivex.internal.operators.single.SingleMap$MapSingleObserver.onSuccess(SingleMap.java:57)
	at io.reactivex.internal.operators.single.SingleJust.subscribeActual(SingleJust.java:30)
	at io.reactivex.Single.subscribe(Single.java:2846)
	at io.reactivex.internal.operators.single.SingleMap.subscribeActual(SingleMap.java:34)
	at io.reactivex.Single.subscribe(Single.java:2846)
	at io.reactivex.internal.operators.single.SingleSubscribeOn$SubscribeOnObserver.run(SingleSubscribeOn.java:89)
	at io.reactivex.Scheduler$DisposeTask.run(Scheduler.java:463)
	at io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:66)
	... 8 more
...Awake
vertxHandler(): Exit
...Done
```

The problem appears to be in io.vertx.core.impl.ContextImpl. (Hence raised here and not on the rx repo).

At lines 273-275, in method `executeBlocking(Handler<Future<T>>, Handler<AsyncResult<T>>, Executor, TaskQueue, PoolMetrics)` there is the following

```
        } catch (Throwable e) {
            res.tryFail(e);
        } finally {

```

However, if the `Handler<AsyncResult<T>> resultHandler` supplied to the method is null, as is the case in the test code below,  then as far as I can tell there is nothing listening to the response of the `Future res` variable. See lines 283-285...

```
        if (resultHandler != null) {
          res.setHandler(ar -> runOnContext(v -> resultHandler.handle(ar)));
        }

```
I think the fix is that the catch block at line 273 should call reportException(e) if resultHandler is null.

```
        } catch (Throwable e) {
          if ( resultHandler != null ) {
            res.tryFail(e);
          } else {
            reportException(e);
          }
        } finally {

```

This is the output with the change suggested above.

```
Starting...
vertxHandler(): Entry
Sleeping...
throwFatal(hello)
...Awake
vertxHandler(): Exit
...Done
Vertx exception handler invoked
java.lang.StackOverflowError: FATAL!!
	at FatalSwallowed.throwFatal(FatalSwallowed.java:71)
	at FatalSwallowed.lambda$3(FatalSwallowed.java:60)
	at io.reactivex.internal.operators.single.SingleMap$MapSingleObserver.onSuccess(SingleMap.java:57)
	at io.reactivex.internal.operators.single.SingleJust.subscribeActual(SingleJust.java:30)
	at io.reactivex.Single.subscribe(Single.java:2846)
	at io.reactivex.internal.operators.single.SingleMap.subscribeActual(SingleMap.java:34)
	at io.reactivex.Single.subscribe(Single.java:2846)
	at io.reactivex.internal.operators.single.SingleSubscribeOn$SubscribeOnObserver.run(SingleSubscribeOn.java:89)
	at io.reactivex.Scheduler$DisposeTask.run(Scheduler.java:463)
	at io.vertx.reactivex.ContextScheduler$ContextWorker$TimedAction.run(ContextScheduler.java:184)
	at io.vertx.reactivex.ContextScheduler$ContextWorker$TimedAction.lambda$execute$1(ContextScheduler.java:169)
	at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$2(ContextImpl.java:272)
	at io.vertx.core.impl.TaskQueue.run(TaskQueue.java:76)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)

```

The test class follows.


```
import io.reactivex.Single;
import io.reactivex.plugins.RxJavaPlugins;
import io.reactivex.schedulers.Schedulers;
import io.vertx.core.AsyncResult;
import io.vertx.core.Handler;
import io.vertx.core.VertxOptions;
import io.vertx.reactivex.SingleHelper;
import io.vertx.reactivex.core.RxHelper;
import io.vertx.reactivex.core.Vertx;

public class FatalSwallowed {

    private final Vertx vertx;

    public static void main(String[] args) {

        System.out.println("Starting...");

        FatalSwallowed fs = new FatalSwallowed();

        // Comment out to see the exception on the console,
       // leave in to see the exception silently ignored
        fs.useVertxSchedulers();

        fs.vertxHandler();

        System.out.println("...Done");
        fs.close();
    }

    public FatalSwallowed() {

        // Long blocking check to allow for debugging
        vertx = Vertx.vertx(new VertxOptions().setBlockedThreadCheckInterval((1000*60)*60));

        vertx.exceptionHandler(t -> {
            System.out.println("Vertx exception handler invoked");
            t.printStackTrace();
        });

        RxJavaPlugins.setErrorHandler( t -> {
            System.out.println("Rx error handler invoked");
            t.printStackTrace();
        });

    }

    private void close() {
        vertx.close();
    }

    private void useVertxSchedulers() {
        RxJavaPlugins.setIoSchedulerHandler(s -> RxHelper.blockingScheduler(vertx));
    }


    private void vertxHandler() {

        System.out.println("vertxHandler(): Entry");
        Handler<AsyncResult<String>> h = getHandler();

        Single.just("hello")
                .map( s -> throwFatal(s) )
                .subscribeOn(Schedulers.io())
                .subscribe(SingleHelper.toObserver(h));

        sleep();
        System.out.println("vertxHandler(): Exit");

    }

    private String throwFatal(String s) {
        System.out.println("throwFatal(" + s + ")");
        throw new StackOverflowError("FATAL!!");
    }

    private void sleep() {

        System.out.println("Sleeping...");

        try {
            Thread.sleep(2000);
        } catch (InterruptedException e) {
            System.out.println("Interrupted");
            e.printStackTrace();
        }

        System.out.println("...Awake");
    }

    public Handler<AsyncResult<String>> getHandler() {

        return new Handler<AsyncResult<String>>() {
            @Override
            public void handle(AsyncResult<String> event) {
                System.out.println("failed=" + event.failed());
                System.out.println("cause=" + event.cause());
                System.out.println("result=" + event.result());
            }

        };
    }
}

```

--- a/src/main/java/io/vertx/core/impl/ContextImpl.java
+++ b/src/main/java/io/vertx/core/impl/ContextImpl.java
@@ -13,16 +13,10 @@ package io.vertx.core.impl;
 
 import io.netty.channel.EventLoop;
 import io.netty.channel.EventLoopGroup;
-import io.vertx.core.AsyncResult;
-import io.vertx.core.Closeable;
-import io.vertx.core.Context;
-import io.vertx.core.Future;
-import io.vertx.core.Handler;
-import io.vertx.core.Promise;
-import io.vertx.core.Vertx;
-import io.vertx.core.json.JsonObject;
+import io.vertx.core.*;
 import io.vertx.core.impl.logging.Logger;
 import io.vertx.core.impl.logging.LoggerFactory;
+import io.vertx.core.json.JsonObject;
 import io.vertx.core.spi.metrics.PoolMetrics;
 import io.vertx.core.spi.tracing.VertxTracer;
 
@@ -181,9 +175,13 @@ abstract class ContextImpl extends AbstractContext {
         if (metrics != null) {
           metrics.end(execMetric, fut.succeeded());
         }
-        if (resultHandler != null) {
-          fut.setHandler(ar -> context.runOnContext(v -> resultHandler.handle(ar)));
-        }
+        fut.setHandler(ar -> {
+          if (resultHandler != null) {
+            context.runOnContext(v -> resultHandler.handle(ar));
+          } else if (ar.failed()) {
+            context.reportException(ar.cause());
+          }
+        });
       };
       Executor exec = workerPool.executor();
       if (queue != null) {

--- a/src/test/java/io/vertx/core/ContextTest.java
+++ b/src/test/java/io/vertx/core/ContextTest.java
@@ -11,11 +11,7 @@
 
 package io.vertx.core;
 
-import io.vertx.core.impl.ContextInternal;
-import io.vertx.core.impl.TaskQueue;
-import io.vertx.core.impl.VertxInternal;
-import io.vertx.core.impl.VertxThread;
-import io.vertx.core.impl.WorkerPool;
+import io.vertx.core.impl.*;
 import io.vertx.test.core.VertxTestBase;
 import org.junit.Test;
 
@@ -314,6 +310,37 @@ public class ContextTest extends VertxTestBase {
         throw failure;
       });
     });
+    await();
+  }
+
+  @Test
+  public void testExceptionInExecutingBlockingWithContextExceptionHandler() {
+    RuntimeException expected = new RuntimeException("test");
+    Context context = vertx.getOrCreateContext();
+    context.exceptionHandler(t -> {
+      assertSame(expected, t);
+      complete();
+    });
+    vertx.exceptionHandler(t -> {
+      fail("Should not be invoked");
+    });
+    context.executeBlocking(promise -> {
+      throw expected;
+    }, null);
+    await();
+  }
+
+  @Test
+  public void testExceptionInExecutingBlockingWithVertxExceptionHandler() {
+    RuntimeException expected = new RuntimeException("test");
+    Context context = vertx.getOrCreateContext();
+    vertx.exceptionHandler(t -> {
+      assertSame(expected, t);
+      complete();
+    });
+    context.executeBlocking(promise -> {
+      throw expected;
+    }, null);
     await();
   }
 

